edb/schema/utils.py

```diff
@@ -438,60 +507,71 @@ def ensure_union_type(
 ) -> Tuple[s_schema.Schema, s_types.Type, bool]:
 
     from edb.schema import objtypes as s_objtypes
+    from edb.schema import types as s_types
 
-    components = set()
+    components: Set[s_types.Type] = set()
     for t in types:
         union_of = t.get_union_of(schema)
         if union_of:
             components.update(union_of.objects(schema))
         else:
             components.add(t)
 
-    components = minimize_class_set_by_most_generic(schema, components)
+    components_list = minimize_class_set_by_most_generic(schema, components)
 
-    if len(components) == 1 and not opaque:
-        return schema, next(iter(components)), False
+    if len(components_list) == 1 and not opaque:
+        return schema, next(iter(components_list)), False
 
     seen_scalars = False
     seen_objtypes = False
     created = False
 
-    for component in components:
+    for component in components_list:
         if component.is_object_type():
             if seen_scalars:
-                raise _union_error(schema, components)
+                raise _union_error(schema, components_list)
             seen_objtypes = True
         else:
             if seen_objtypes:
-                raise _union_error(schema, components)
+                raise _union_error(schema, components_list)
             seen_scalars = True
 
     if seen_scalars:
-        uniontype = components[0]
-        for t1 in components[1:]:
-            uniontype = uniontype.find_common_implicitly_castable_type(
-                t1, schema)
+        uniontype: s_types.Type = components_list[0]
+        for t1 in components_list[1:]:
+
+            common_type = uniontype.\
+                find_common_implicitly_castable_type(t1, schema)
 
-        if uniontype is None:
-            raise _union_error(schema, components)
+            if common_type is None:
+                raise _union_error(schema, components_list)
+            else:
+                uniontype = common_type
     else:
         schema, uniontype, created = s_objtypes.get_or_create_union_type(
-            schema, components=components, opaque=opaque, module=module)
+            schema,
+            components=cast(Iterable[s_objtypes.ObjectType],
```

Comment:
It's a bit weird that you had to put a cast here. What type is `components_list` without a cast?

Reply:
Right :+1: